peerdid/peer_did_utils.py

```diff
@@ -0,0 +1,201 @@
+    except TypeError:
+        raise ValueError('incorrect varint provided')
+
+
+def _add_prefix(key_type: Union[PublicKeyTypeAgreement, PublicKeyTypeAuthentication], data: bytes) -> bytes:
+    """
+    Adds prefix to a data
+    :param key_type: type of key
+    :param data: data to be prefixed
+    :return: prefixed data
+    """
+    prefix = varint.encode(key_type.value)
+    return b''.join([prefix, data])
+
+
+def encode_filename(filename: str) -> str:
+    """
+    Encodes filename to SHA256 string
+    :param filename: name of file
+    :return: encoded filename as SHA256 string
+    """
+    return hashlib.sha256(filename.encode()).hexdigest()
+
+
+def _build_did_doc_numalgo_0(peer_did: PEER_DID) -> dict:
+    """
+    Helper method to create did_doc according to numalgo 0
+    :param peer_did: peer_did to resolve
+    :raises ValueError: if peer_did contains encryption key instead of signing
+    :return: did_doc
+    """
+    inception_key = peer_did[11:]
+    decoded_encnumbasis = _decode_encnumbasis(inception_key, peer_did)
+    if not decoded_encnumbasis['type'] in PublicKeyTypeAuthentication.__members__:
+        raise ValueError('Invalid key type (encryption instead of signing)')
+    did_doc = {
+        'id': peer_did,
+        'authentication': decoded_encnumbasis
+    }
+    return did_doc
+
+
+def _build_did_doc_numalgo_2(peer_did: PEER_DID) -> dict:
+    """
+    Helper method to create did_doc according to numalgo 2
+    :param peer_did: peer_did to resolve
+    :return: did_doc
+    """
+    keys = peer_did[11:]
+    keys = keys.split('.')
+    services = []
+    keys_without_purpose_code = []
+    for key in keys:
+        if key[0] != 'S':
+            keys_without_purpose_code.append(key[2:])
+        else:
+            services.append(key[1:])
+    decoded_encnumbasises = [_decode_encnumbasis(key, peer_did) for key in keys_without_purpose_code]
+    decoded_services = [_decode_service(service, peer_did) for service in services]
+
+    authentication = []
+    key_agreement = []
+    for i in range(len(decoded_encnumbasises)):
+        if decoded_encnumbasises[i]['type'] in PublicKeyTypeAuthentication.__members__ and keys[i][0] == 'V':
+            authentication.append(decoded_encnumbasises[i])
+        elif decoded_encnumbasises[i]['type'] in PublicKeyTypeAgreement.__members__ and keys[i][0] == 'E':
+            key_agreement.append(decoded_encnumbasises[i])
+        else:
+            raise ValueError(f'Invalid key type of: {keys[i]}')
+    did_doc = {
+        'id': peer_did,
+        'authentication': authentication,
+        'keyAgreement': key_agreement,
+        'service': decoded_services
+    }
+    return did_doc
+
+
+def _check_if_base58_encoded(key: str) -> bool:
+    """
+    Checks if key base58 encoded
+    :param key: any string
+    :return: true if key base58 encoded, otherwise false
+    """
+    base58_pattern = re.compile(r"^[123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz]+$")
```

Comment:
I'm not sure this check is enough.
I think we may have to do the real decoding, and check if the decoded length matches the expected key length.
See, for example, [link](https://github.com/hyperledger/indy-plenum/blob/f812a72197bc0740d4c6b75b796beff8f88c73f9/plenum/common/messages/fields.py#L354)